src/Elastic.Apm.SqlClient/SqlEventListener.cs

```diff
@@ -0,0 +1,139 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using Elastic.Apm.Api;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model;
+
+namespace Elastic.Apm.SqlClient
+{
+	internal class SqlEventListener : EventListener
+	{
+		private readonly ApmAgent _apmAgent;
+		private readonly IApmLogger _logger;
+
+		private readonly ConcurrentDictionary<int, (Span Span, long Start)> _spans = new ConcurrentDictionary<int, (Span, long)>();
+
+		public SqlEventListener(IApmAgent apmAgent)
+		{
+			_apmAgent = (ApmAgent)apmAgent;
+			_logger = _apmAgent.Logger.Scoped(nameof(SqlEventListener));
+		}
+
+		protected override void OnEventSourceCreated(EventSource eventSource)
+		{
+			if (eventSource != null && eventSource.Name == "Microsoft-AdoNet-SystemData")
+				EnableEvents(eventSource, EventLevel.Informational, (EventKeywords)1);
+
+			base.OnEventSourceCreated(eventSource);
+		}
+
+		private const int BeginExecuteEventId = 1;
+		private const int EndExecuteId = 2;
+
+		protected override void OnEventWritten(EventWrittenEventArgs eventData)
+		{
+			if (eventData?.Payload == null) return;
+
+			try
+			{
+				switch (eventData.EventId)
+				{
+					case BeginExecuteEventId:
+						ProcessBeginExecute(eventData.Payload);
+						break;
+					case EndExecuteId:
+						ProcessEndExecute(eventData.Payload);
+						break;
+				}
+			}
+			catch (Exception ex)
+			{
+				_logger?.Error()?.LogException(ex, "Error has occurred during handle event from SqlClient. EventData: {EventData}", eventData);
+			}
+		}
+
+		private void ProcessBeginExecute(IReadOnlyList<object> payload)
+		{
+			if (payload.Count != 4)
+			{
+				_logger?.Debug()
+					?.Log("BeginExecute event has {PayloadCount} payload items instead of 4. Event processing is skipped.", payload.Count);
+				return;
+			}
+
+			var id = Convert.ToInt32(payload[0]);
+			var dataSource = Convert.ToString(payload[1]);
+			var database = Convert.ToString(payload[2]);
+			var commandText = Convert.ToString(payload[3]);
+			var start = Stopwatch.GetTimestamp();
+
+			_logger?.Trace()
+				?.Log("Process BeginExecute event. Id: {Id}. Data source: {DataSource}. Database: {Database}. CommandText: {CommandText}.", id,
+					dataSource, database, commandText);
+
+			var spanName = !string.IsNullOrWhiteSpace(commandText)
```

Comment:
Nice, as said [here](https://github.com/elastic/apm-agent-dotnet/pull/704#pullrequestreview-360691579) I think we'll have the `commandText` with newer versions - so over time this'll hopefully capture the CommandText more often.

Reply:
The only one question is left: is it enough to use database name in case of command text is missing? Maybe we need somehow combine data source with database name to span name?

Reply:
This could be relevant for searching and maybe grouping - but given that we won't have any dynamic data here I think it would not make a big difference if we add the database source. I'd keep it as it is.